bundle/edu.gemini.programexport.servlet/src/main/scala/edu/gemini/programexport/servlet/ProgramExportServlet.scala

```diff
@@ -0,0 +1,209 @@
+      case SPComponentType.INFO_PROGRAMNOTE => //| SPComponentType.INFO_SCHEDNOTE | SPComponentType.INFO_NOTE =>
+        processProgramNote(n.getDataObject.asInstanceOf[SPNote])
+    }
+
+  // Program-specific notes: must go through the components and check type.
+  // The obs components for all of these types are SPNotes or subclasses.
+  def processProgramNote(note: SPNote): String = {
+    val title = note.getTitle
+    val text  = note.getNote
+    ""
+  }
+
+  def processGroup(ispGroup: ISPGroup): String = {
+    val spGroup = ispGroup.getDataObject.asInstanceOf[SPGroup]
+    val key = ispGroup.getNodeKey
+    val name = spGroup.getTitle
+
+    // Get a list of observation node keys for the scheduling group
+    val observationIds = combine(ispGroup.getObservations.asScala.map {
+      _.getNodeKey.toString
+    }.toList)
+
+    ""
+  }
+
+  def processObservation(ispObservation: ISPObservation): String = {
+    val spObservation = ispObservation.getDataObject.asInstanceOf[SPObservation]
+    val id = ispObservation.getNodeKey
+    val title = spObservation.getTitle
+    val phase2Status = spObservation.getPhase2Status.displayValue
+
+    // TODO: If this is None, which it could be, do we want to display it?
+    val execStatusOverride = spObservation.getExecStatusOverride.asScalaOpt.map(_.displayValue)
+
+    val priority = spObservation.getPriority.displayValue
+    val tooOverrideRapid = spObservation.isOverrideRapidToo.toYesNo.displayValue
+
+    // TODO: Don't know if this is correct way to handle this.
+    val setupTimeType = spObservation.getSetupTimeType.toString
+
+    // TODO: Also need acq overhead. Where is this? In setup time type?
+
+    // Site quality is an obscomponent but not wrapped in an ISP object.
+    // Is this the proper way to get at it? Otherwise we'll have to match on
+    // SPComponentType instead of ISPNode, I think.
+    val conditions = SPTreeUtil.findObsCondNode(ispObservation).getDataObject.asInstanceOf[SPSiteQuality]
+    val cc = conditions.getCloudCover.displayValue
+    val sb = conditions.getSkyBackground.displayValue
+    val iq = conditions.getImageQuality.displayValue
+    val wv = conditions.getWaterVapor.displayValue
+
+    val ecType = conditions.getElevationConstraintType.displayValue
+    val ecMin = conditions.getElevationConstraintMin
+    val ecMax = conditions.getElevationConstraintMax
+
+    val timingWindows = combine(conditions.getTimingWindows.asScala.map { tw =>
+      // TODO: in ms. Is this what we want?
+      val start = tw.getStart
+      val duration = tw.getDuration
+      val period = tw.getPeriod
+
+      val repeat = tw.getRepeat
+      ""
+    }.toList)
+
+    // TODO: Now I need a TargetEnvironment... Do I go through ObsContext?
+
+    ""
+  }
+}
+
+object ProgramExportServlet {
+  val Log: Logger = Logger.getLogger(getClass.getName)
+
+  val IdParam: String = "id"
+  case class ParamParser(req: HttpServletRequest) {
+    def id: Throwable \/ SPProgramID =
```

Comment:
Being this new code, should we try to use cats rather than scalaz?

Reply:
Ordinarily yes but I think it would be best to not mix the two and scalaz is already prevalent in the old codebase.

Reply:
I definitely prefer cats to scalaz, but it is as Shane said.